test/kubernetes_unittest.cc

```diff
@@ -68,4 +89,142 @@ TEST_F(KubernetesTest, GetNodeMetadata) {
+  json::value empty_cluster = json::object({
+    {"blobs", json::object({
+      {"services", json::array({})},
+    })},
+  });
+  EXPECT_EQ(empty_cluster->ToString(), m.metadata().metadata->ToString());
+}
+
+TEST_F(KubernetesTest, GetClusterMetadataEmptyService) {
+  Configuration config(std::stringstream(
+    "KubernetesClusterName: TestClusterName\n"
+    "KubernetesClusterLocation: TestClusterLocation\n"
+    "MetadataIngestionRawContentVersion: TestVersion\n"
+  ));
+  Environment environment(config);
+  json::value service = json::object({
+    {"metadata", json::object({
+      {"name", json::string("testname")},
+      {"namespace", json::string("testnamespace")},
+    })}
+  });
+  KubernetesReader reader(config, nullptr);  // Don't need HealthChecker.
+  UpdateServiceToMetadataCache(
+      reader, service->As<json::Object>(), /*is_deleted=*/false);
+  const auto m = GetClusterMetadata(reader, Timestamp());
+  EXPECT_EQ(0, m.ids().size());
+  EXPECT_EQ(MonitoredResource("k8s_cluster", {
+    {"cluster_name", "TestClusterName"},
+    {"location", "TestClusterLocation"},
+  }), m.resource());
+  EXPECT_EQ("TestVersion", m.metadata().version);
+  EXPECT_EQ(false, m.metadata().is_deleted);
+  EXPECT_EQ(Timestamp(), m.metadata().created_at);
+  EXPECT_EQ(Timestamp(), m.metadata().collected_at);
+  json::value expected_cluster = json::object({
+    {"blobs", json::object({
+      {"services", json::array({
+        json::object({
+          {"api", json::object({
+            {"pods", json::array({})},
+            {"raw", std::move(service)},
+            {"version", json::string("1.6")},  // Hard-coded in kubernetes.cc.
+          })},
+        }),
+      })},
+    })},
+  });
+  EXPECT_EQ(expected_cluster->ToString(), m.metadata().metadata->ToString());
+}
+
+TEST_F(KubernetesTest, GetClusterMetadataServiceWithPods) {
+  Configuration config(std::stringstream(
+    "KubernetesClusterName: TestClusterName\n"
+    "KubernetesClusterLocation: TestClusterLocation\n"
+    "MetadataIngestionRawContentVersion: TestVersion\n"
+  ));
+  Environment environment(config);
+  json::value service = json::object({
+    {"metadata", json::object({
+      {"name", json::string("testname")},
+      {"namespace", json::string("testnamespace")},
+    })}
+  });
+  json::value endpoints = json::object({
+    {"metadata", json::object({
+      {"name", json::string("testname")},
+      {"namespace", json::string("testnamespace")},
+    })},
+    {"subsets", json::array({
+      json::object({
+        {"addresses", json::array({
+          json::object({
+            {"targetRef", json::object({
+              {"kind", json::string("Pod")},
+              {"name", json::string("my-pod")},
+            })},
+          }),
+        })},
+      }),
+    })},
+  });
+  KubernetesReader reader(config, nullptr);  // Don't need HealthChecker.
+  UpdateServiceToMetadataCache(
+      reader, service->As<json::Object>(), /*is_deleted=*/false);
```

Comment:
Do we need to also test deleted services in a separate test?

Reply:
added a separate test.